hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,324 @@
+//! [ascent] datalog implementation of analysis.
+//! Since ascent-(macro-)generated code generates a bunch of warnings,
+//! keep code in here to a minimum.
+#![allow(
+    clippy::clone_on_copy,
+    clippy::unused_enumerate_index,
+    clippy::collapsible_if
+)]
+
+use ascent::lattice::{BoundedLattice, Lattice};
+use itertools::zip_eq;
+use std::cmp::Ordering;
+use std::hash::Hash;
+use std::ops::{Index, IndexMut};
+
+use hugr_core::extension::prelude::{MakeTuple, UnpackTuple};
+use hugr_core::ops::OpType;
+use hugr_core::types::Signature;
+use hugr_core::{Hugr, HugrView, IncomingPort, Node, OutgoingPort, PortIndex as _};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum IO {
+    Input,
+    Output,
+}
+
+/// Clients of the dataflow framework (particular analyses, such as constant folding)
+/// must implement this trait (including providing an appropriate domain type `PV`).
+pub trait DFContext<PV: AbstractValue>: Clone + Eq + Hash + std::ops::Deref<Target = Hugr> {
```

Comment:
`Deref<Hugr>` is semantically equivalent to your `fn hugr(&self) -> &Hugr` but syntactically much neater - you can do just `*x` rather than `x.hugr()` and you can directly call `x.foo()` rather than `x.hugr().foo()`.

So I quite like it, but I haven't understood the objection yet. (IIUC the "proliferation of generic parameters" applies only if we return an arbitrary `HugrView`? So we could return `impl HugrView`, would that avoid the proliferation? Then, what, this change would allow running the dataflow analysis on a sub-hugr (via a RegionView) with certainty that the analysis wasn't touching anything outside that? (I guess that could be good...)

Reply:
Hmmmm. Just possibly this would enable us to interpret leaf-op `Call`s by firing up a separate analysis, finishing that, and returning results into the parent analysis. So that might allow that as a way to resolve the Call question. (But you could always just make a new Hugr rather than using a view - the cost of copying nodes seems small compared to running the analysis.) And I don't think this generalizes to also beiing able to do a context-*insensitive* analysis (where all leaf-`Call`s share variables), so probably isn't a good technique overall.

Reply:
Now that we're using `ascent_run` I've separated DFContext from HugrView altogether so I think this is addressed.